### DISCUSSION TITLE
Cannot test WPF library using `net9.0-windows` target framework

### DISCUSSION THREAD

**Discussion by SetTrend** [microsoft/testfx#4107] [Q&A] [answered]
I just created a new SDK test project using Visual Studio 2022 v17.12.1 and the new .NET 9 SDK for testing my WPF library.

It seems that I'm not able to select `net9.0-windows` as the target framework. I'm getting this error message in the Errors panel:

```
Project '..\..\Converter.Html\Converter.Html.csproj' targets 'net9.0-windows'. It cannot be referenced
by a project that targets '.NETCoreApp,Version=v9.0'.
```

Manually changing the target framework to `net9.0-windows` doesn't work. It yields the following error message:

```
Assets file 'D:\Documents\Repos\WPF\Converter\+Tests\Html.Tests\obj\project.assets.json' doesn't have
a target for 'net9.0-windows'. Ensure that restore has run and that you have included 'net9.0-windows'
in the TargetFrameworks for your project.
```

<br/>

FWIW: This is my test project file:

```xml
<Project Sdk="MSTest.Sdk/3.6.1">

  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
    <LangVersion>latest</LangVersion>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <ProduceReferenceAssembly>False</ProduceReferenceAssembly>
    <IsPublishable>False</IsPublishable>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <DebugType>none</DebugType>
    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\Converter.Html\Converter.Html\Converter.Html.csproj" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="Microsoft.NET.Test.Sdk" Version="17.12.0" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="Microsoft.Testing.Extensions.CodeCoverage" Version="17.12.6" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="Microsoft.Testing.Extensions.TrxReport" Version="1.4.3" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="MSTest.Analyzers" Version="3.6.3">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="MSTest.TestAdapter" Version="3.6.3" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="MSTest.TestFramework" Version="3.6.3" />
  </ItemGroup>

</Project>
```

**Comment by SetTrend** [comment]
Hmm … I manually changed the target framework of my unit test project to `net9.0-windows`. Now I can compile, but NuGet claims that the test packages do have vulnerabilities.

An no matter how often I update the test NuGet packages now … As soon as I reload the project, I'm getting displayed the same vulnerability report again, and none of the NuGet packages seem to have been updated. Apparently the NuGet packages don't seem to get updated:

![NuGet updates](https://github.com/user-attachments/assets/e6b1d000-1fbc-4692-b243-bacb9dcc4034)

Moreover, `net9.0-windows` is not listed as an available .NET target framework.

**Reply by nohwnd** [reply, marked as answer]
Yes this is a problem with the SDK style project, you cannot update via the nuget manager. On the top of your project it says:

```
<Project Sdk="MSTest.Sdk/3.6.1">
```
Change the version there to 3.6.3.  And also delete the explicit references to packages. 

 Or  if you don't wish to use the MSTest SDK style project, revert the top line to say 

```
<Project Sdk="Microsoft.NET.Sdk">
```

and add this to the property group:

```
    <EnableMSTestRunner>true</EnableMSTestRunner>
    <OutputType>Exe</OutputType>
```

This should allow you to keep your project as is,  it looks like it already has all the other needed props and items.

**Reply by Evangelink** [reply]
Adding up onto what @nohwnd mentioned.

1. When using `MSTest.Sdk` the NuGet integration is currently limited. I have created https://github.com/microsoft/testfx/issues/4114 for us to improve documentation in that regard. Note that Dependabot is handling version updates when using the `global.json`.

1. I recommend you to have a look at the [migrating to MSTest Sdk documentation](https://learn.microsoft.com/dotnet/core/testing/unit-testing-mstest-sdk#migrate-to-mstest-sdk) and more specifically the simplification part as you want your project to not explicitely reference the packages but instead leave the handling to the SDK.

Your project should look like:

```xml
<Project Sdk="MSTest.Sdk/3.6.3">

  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
    <LangVersion>latest</LangVersion>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <ProduceReferenceAssembly>False</ProduceReferenceAssembly>
    <IsPublishable>False</IsPublishable>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <DebugType>none</DebugType>
    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\Converter.Html\Converter.Html\Converter.Html.csproj" />
  </ItemGroup>

</Project>
```

1. I have just done a simple test and I am correctly able to build and run tests for `net9.0-windows` 
![image](https://github.com/user-attachments/assets/0eb202bb-dc67-49ee-b8be-d4d4cb9286b8)

If you are having some issue there, could you please create a small reproducer and share it with us?

**Reply by SetTrend** [reply]
Thank you for your very valuable help!

So, you suggest the project template item shipped with VS 2022 isn't valid and up-to-date?

@Evangelink:

I didn't do anything special. I just created a test project using the GUI.

Then, when updating the NuGet packages, NuGet Package Manager added the explicit references:
_(please click "Details" to see the screenshots:)_

<details>

![1](https://github.com/user-attachments/assets/c63d47b5-808b-4dff-bcb7-2c97c79a87d1)
![2](https://github.com/user-attachments/assets/1f470ab1-b642-4a89-ad99-17a07a2b9ee6)
![3](https://github.com/user-attachments/assets/3b133c4f-6cc2-4ef2-8151-c3b55332e074)
![4](https://github.com/user-attachments/assets/cdbf34d6-9106-4c5f-a59c-9d15bf1b9176)
![5](https://github.com/user-attachments/assets/9e87ced2-3afc-489a-a873-07ae9b0bb4dc)

</details>

<br/>

I marked this discussion as being answered. Yet, shouldn't this be reported somewhere else than to the documentation team?

**Reply by Evangelink** [reply]
> So, you suggest the project template item shipped with VS 2022 isn't valid and up-to-date?

We are shipping templates within .NET SDKs and our latest insertion is linked to .NET 9 GA. When we found about the security on the deps, it was already too late to do the change for the current SDK. The new version will be shipped with the next patch version of SDK.
